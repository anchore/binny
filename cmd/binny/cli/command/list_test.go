package command

import (
	"errors"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type renderListTableTest struct {
	name     string
	statuses []toolStatus
}

func renderListTableTestCases() []renderListTableTest {
	return []renderListTableTest{
		{
			name:     "empty",
			statuses: []toolStatus{},
		},
		{
			name: "has update",
			statuses: []toolStatus{
				{
					Name:             "syft",
					WantVersion:      "latest",
					ResolvedVersion:  "v1.0.0",
					InstalledVersion: "v0.105.1",
					Constraint:       "<= v1.0.0",
					IsInstalled:      true,
					HashIsValid:      true,
					Error:            nil,
				},
			},
		},
		{
			name: "invalid hash",
			statuses: []toolStatus{
				{
					Name:             "syft",
					WantVersion:      "v0.105.1",
					ResolvedVersion:  "v0.105.1",
					InstalledVersion: "v0.105.1",
					Constraint:       "<= v1.0.0",
					IsInstalled:      true,
					HashIsValid:      false,
					Error:            nil,
				},
			},
		},
		{
			name: "error",
			statuses: []toolStatus{
				{
					Name:             "syft",
					WantVersion:      "latest",
					ResolvedVersion:  "v1.0.0",
					InstalledVersion: "v1.0.0",
					Constraint:       "<= v1.0.0",
					IsInstalled:      true,
					HashIsValid:      false,
					Error:            errors.New("something is wrong"),
				},
			},
		},
		{
			name: "unknown wanted version",
			statuses: []toolStatus{
				{
					Name:             "syft",
					WantVersion:      "?",
					ResolvedVersion:  "v1.0.1",
					InstalledVersion: "v1.0.0",
					Constraint:       "<= v1.0.1",
					IsInstalled:      true,
					HashIsValid:      false,
					Error:            nil,
				},
			},
		},
		{
			name: "not installed",
			statuses: []toolStatus{
				{
					Name:             "syft",
					WantVersion:      "latest",
					ResolvedVersion:  "v1.0.0",
					InstalledVersion: "",
					Constraint:       "<= v1.0.0",
					IsInstalled:      false,
					HashIsValid:      true,
					Error:            nil,
				},
			},
		},
		{
			name: "no update",
			statuses: []toolStatus{
				{
					Name:             "syft",
					WantVersion:      "latest",
					ResolvedVersion:  "v1.0.0",
					InstalledVersion: "v1.0.0",
					Constraint:       "<= v1.0.0",
					IsInstalled:      true,
					HashIsValid:      true,
					Error:            nil,
				},
			},
		},
		{
			name: "sort by name",
			statuses: []toolStatus{
				{
					Name:             "syft",
					WantVersion:      "latest",
					ResolvedVersion:  "v1.0.0",
					InstalledVersion: "v0.105.1",
					Constraint:       "<= v1.0.0",
					IsInstalled:      true,
					HashIsValid:      true,
					Error:            nil,
				},
				{
					Name:             "grype",
					WantVersion:      "v0.74.0",
					ResolvedVersion:  "v0.74.0",
					InstalledVersion: "v0.53.0",
					Constraint:       "<= v1.0.0",
					IsInstalled:      true,
					HashIsValid:      true,
					Error:            nil,
				},
			},
		},
	}
}

func Test_renderListTable(t *testing.T) {

	for _, tt := range renderListTableTestCases() {
		t.Run(tt.name, func(t *testing.T) {
			got := renderListTable(tt.statuses)
			snaps.MatchSnapshot(t, got)
		})
	}
}

func Test_renderListUpdatesTable(t *testing.T) {

	for _, tt := range renderListTableTestCases() {
		t.Run(tt.name, func(t *testing.T) {
			got := renderListUpdatesTable(tt.statuses)
			snaps.MatchSnapshot(t, got)
		})
	}
}

func Test_renderListJSON(t *testing.T) {

	t.Run("updates", func(t *testing.T) {
		for _, tt := range renderListTableTestCases() {
			t.Run(tt.name, func(t *testing.T) {
				got, err := renderListJSON(tt.statuses, true, "")
				require.NoError(t, err)
				snaps.MatchSnapshot(t, got)
			})
		}
	})

	t.Run("no updates", func(t *testing.T) {
		for _, tt := range renderListTableTestCases() {
			t.Run(tt.name, func(t *testing.T) {
				got, err := renderListJSON(tt.statuses, false, "")
				require.NoError(t, err)
				snaps.MatchSnapshot(t, got)
			})
		}
	})

	t.Run("jq", func(t *testing.T) {
		testStatuses := []toolStatus{
			{
				Name:             "syft",
				WantVersion:      "latest",
				ResolvedVersion:  "v1.0.0",
				InstalledVersion: "v0.105.1",
				Constraint:       "<= v1.0.0",
				IsInstalled:      true,
				HashIsValid:      true,
				Error:            nil,
			},
			{
				Name:             "grype",
				WantVersion:      "v0.74.0",
				ResolvedVersion:  "v0.74.0",
				InstalledVersion: "v0.53.0",
				Constraint:       "<= v1.0.0",
				IsInstalled:      true,
				HashIsValid:      true,
				Error:            nil,
			},
		}

		tests := []struct {
			name     string
			statuses []toolStatus
			jq       string
			wantErr  require.ErrorAssertionFunc
		}{
			{
				name:     "show keys",
				statuses: testStatuses,
				jq:       "keys",
			},
			{
				name:     "bad jq expression",
				statuses: testStatuses,
				jq:       "shdjfkshf",
				wantErr:  require.Error,
			},
			{
				name:     "filter by name",
				statuses: testStatuses,
				jq:       ".tools[] | select(.name == \"syft\")",
			},
			{
				name:     "raw scalar values",
				statuses: testStatuses,
				jq:       ".tools[].wantVersion",
			},
		}
		for _, tt := range tests {
			t.Run(tt.name, func(t *testing.T) {
				if tt.wantErr == nil {
					tt.wantErr = require.NoError
				}
				got, err := renderListJSON(tt.statuses, false, tt.jq)
				tt.wantErr(t, err)
				snaps.MatchSnapshot(t, got)
			})
		}
	})

}

func Test_summarizeVersion(t *testing.T) {

	tests := []struct {
		name string
		v    string
		want string
	}{
		{
			name: "empty",
			v:    "",
			want: "",
		},
		{
			name: "semver",
			v:    "v1.0.0",
			want: "v1.0.0",
		},
		{
			name: "commit-sha",
			v:    "250ca084c8ffd28fad8bf9d8725e2b0d5b8b11e2",
			want: "250ca08",
		},
		{
			name: "40-char non-sha",
			v:    "SOMETHING-ffd28fad8bf9d8725e2b0d5b8b11e2",
			want: "SOMETHING-ffd28fad8bf9d8725e2b0d5b8b11e2",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equal(t, tt.want, summarizeGitVersion(tt.v))
		})
	}
}
