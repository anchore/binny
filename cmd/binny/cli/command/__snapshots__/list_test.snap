
[Test_renderListTable/empty - 1]
no tools configured or installed
---

[Test_renderListTable/has_update - 1]
 TOOL  DESIRED VERSION  CONSTRAINT                                                                        
──────────────────────────────────────────────────────────────────────────────────────────────────────────
 syft  latest (v1.0.0)  <= v1.0.0   installed version (v0.105.1) does not match resolved version (v1.0.0) 
---

[Test_renderListTable/invalid_hash - 1]
 TOOL  DESIRED VERSION  CONSTRAINT                  
────────────────────────────────────────────────────
 syft  v0.105.1         <= v1.0.0   hash is invalid 
---

[Test_renderListTable/error - 1]
 TOOL  DESIRED VERSION  CONSTRAINT                     
───────────────────────────────────────────────────────
 syft  latest (v1.0.0)  <= v1.0.0   something is wrong 
---

[Test_renderListTable/unknown_wanted_version - 1]
 TOOL  DESIRED VERSION  CONSTRAINT                         
───────────────────────────────────────────────────────────
 syft  ? (v1.0.1)       <= v1.0.1   tool is not configured 
---

[Test_renderListTable/not_installed - 1]
 TOOL  DESIRED VERSION  CONSTRAINT                
──────────────────────────────────────────────────
 syft  latest (v1.0.0)  <= v1.0.0   not installed 
---

[Test_renderListTable/no_update - 1]
 TOOL  DESIRED VERSION  CONSTRAINT   
─────────────────────────────────────
 syft  latest (v1.0.0)  <= v1.0.0    
---

[Test_renderListTable/sort_by_name - 1]
 TOOL   DESIRED VERSION  CONSTRAINT                                                                        
───────────────────────────────────────────────────────────────────────────────────────────────────────────
 grype  v0.74.0          <= v1.0.0   installed version (v0.53.0) does not match resolved version (v0.74.0) 
 syft   latest (v1.0.0)  <= v1.0.0   installed version (v0.105.1) does not match resolved version (v1.0.0) 
---

[Test_renderListUpdatesTable/empty - 1]
no tools to check
---

[Test_renderListUpdatesTable/has_update - 1]
 TOOL  UPDATE            
─────────────────────────
 syft  v0.105.1 → v1.0.0 
---

[Test_renderListUpdatesTable/invalid_hash - 1]
all tools up to date
---

[Test_renderListUpdatesTable/error - 1]
 TOOL  UPDATE             
──────────────────────────
 syft  something is wrong 
---

[Test_renderListUpdatesTable/unknown_wanted_version - 1]
all tools up to date
---

[Test_renderListUpdatesTable/not_installed - 1]
 TOOL  UPDATE        
─────────────────────
 syft  not installed 
---

[Test_renderListUpdatesTable/no_update - 1]
all tools up to date
---

[Test_renderListUpdatesTable/sort_by_name - 1]
 TOOL   UPDATE            
──────────────────────────
 grype  v0.53.0 → v0.74.0 
 syft   v0.105.1 → v1.0.0 
---

[Test_renderListJSON/updates/empty - 1]
{
  "tools": []
}

---

[Test_renderListJSON/updates/has_update - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "v1.0.0",
      "installedVersion": "v0.105.1",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": true
    }
  ]
}

---

[Test_renderListJSON/updates/invalid_hash - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "v0.105.1",
      "resolvedVersion": "v0.105.1",
      "installedVersion": "v0.105.1",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": false
    }
  ]
}

---

[Test_renderListJSON/updates/error - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "",
      "installedVersion": "",
      "constraint": "",
      "isInstalled": true,
      "hashIsValid": false,
      "error": {}
    }
  ]
}

---

[Test_renderListJSON/updates/unknown_wanted_version - 1]
{
  "tools": []
}

---

[Test_renderListJSON/updates/not_installed - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "v1.0.0",
      "installedVersion": "",
      "constraint": "<= v1.0.0",
      "isInstalled": false,
      "hashIsValid": true
    }
  ]
}

---

[Test_renderListJSON/updates/no_update - 1]
{
  "tools": []
}

---

[Test_renderListJSON/updates/sort_by_name - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "v1.0.0",
      "installedVersion": "v0.105.1",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": true
    },
    {
      "name": "grype",
      "wantVersion": "v0.74.0",
      "resolvedVersion": "v0.74.0",
      "installedVersion": "v0.53.0",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": true
    }
  ]
}

---

[Test_renderListJSON/no_updates/empty - 1]
{
  "tools": []
}

---

[Test_renderListJSON/no_updates/has_update - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "v1.0.0",
      "installedVersion": "v0.105.1",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": true
    }
  ]
}

---

[Test_renderListJSON/no_updates/invalid_hash - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "v0.105.1",
      "resolvedVersion": "v0.105.1",
      "installedVersion": "v0.105.1",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": false
    }
  ]
}

---

[Test_renderListJSON/no_updates/error - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "v1.0.0",
      "installedVersion": "v1.0.0",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": false,
      "error": {}
    }
  ]
}

---

[Test_renderListJSON/no_updates/unknown_wanted_version - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "?",
      "resolvedVersion": "v1.0.1",
      "installedVersion": "v1.0.0",
      "constraint": "<= v1.0.1",
      "isInstalled": true,
      "hashIsValid": false
    }
  ]
}

---

[Test_renderListJSON/no_updates/not_installed - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "v1.0.0",
      "installedVersion": "",
      "constraint": "<= v1.0.0",
      "isInstalled": false,
      "hashIsValid": true
    }
  ]
}

---

[Test_renderListJSON/no_updates/no_update - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "v1.0.0",
      "installedVersion": "v1.0.0",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": true
    }
  ]
}

---

[Test_renderListJSON/no_updates/sort_by_name - 1]
{
  "tools": [
    {
      "name": "syft",
      "wantVersion": "latest",
      "resolvedVersion": "v1.0.0",
      "installedVersion": "v0.105.1",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": true
    },
    {
      "name": "grype",
      "wantVersion": "v0.74.0",
      "resolvedVersion": "v0.74.0",
      "installedVersion": "v0.53.0",
      "constraint": "<= v1.0.0",
      "isInstalled": true,
      "hashIsValid": true
    }
  ]
}

---

[Test_renderListJSON/jq/show_keys - 1]
[
  "tools"
]

---

[Test_renderListJSON/jq/bad_jq_expression - 1]

---

[Test_renderListJSON/jq/filter_by_name - 1]
{
  "constraint": "<= v1.0.0",
  "hashIsValid": true,
  "installedVersion": "v0.105.1",
  "isInstalled": true,
  "name": "syft",
  "resolvedVersion": "v1.0.0",
  "wantVersion": "latest"
}

---

[Test_renderListJSON/jq/raw_scalar_values - 1]
latest
v0.74.0

---
